wasm/derive/src/validator/token.rs

```diff
@@ -0,0 +1,179 @@
+                    ::iroha_wasm::data_model::prelude::DoesAccountHavePermissionToken {
+                        account_id: account_id.clone().into(),
+                        permission_token,
+                    }
+                )
+                .execute()
+                .try_into()
+                .dbg_expect("Failed to convert `DoesAccountHavePermission` query result into `bool`")
+            }
+        }
+    }
+}
+
+#[allow(clippy::arithmetic_side_effects)] // Triggers on quote! side
+fn impl_try_from_permission_token(
+    impl_generics: &syn::ImplGenerics<'_>,
+    ty_generics: &syn::TypeGenerics<'_>,
+    where_clause: Option<&syn::WhereClause>,
+    ident: &syn::Ident,
+    fields: &syn::punctuated::Punctuated<syn::Field, syn::token::Comma>,
+) -> proc_macro2::TokenStream {
+    let field_initializers = fields.iter().map(|field| {
+        let field_ident = field.ident.as_ref().expect("Field must have an identifier");
+        let field_literal = proc_macro2::Literal::string(&field_ident.to_string());
+        let syn::Type::Path(field_type) = &field.ty else {
+            panic!("Field must have a type path");
+        };
+
+        let code = quote! {
+            #field_ident: <
+                #field_type
+                as
+                ::core::convert::TryFrom<::iroha_wasm::data_model::prelude::Value>
+            >::try_from(token
+                .param(&::iroha_wasm::parse!(#field_literal as ::iroha_wasm::data_model::prelude::Name))
+                .ok_or(
+                    ::iroha_wasm::validator::PermissionTokenConversionError::Param(#field_literal)
+                )?
+                .clone()
+            )
+            .map_err(|err| {
+                ::iroha_wasm::validator::PermissionTokenConversionError::Value(
+                    ::alloc::string::ToString::to_string(&err)
+                )
+            })?
+        };
+        code
+    });
+
+    quote! {
+        impl #impl_generics ::core::convert::TryFrom<::iroha_wasm::data_model::permission::Token> for #ident #ty_generics
+        #where_clause
+        {
+            type Error = ::iroha_wasm::validator::PermissionTokenConversionError;
+
+            #[allow(unused)] // `params` can be unused if token has none
+            fn try_from(
+                token: ::iroha_wasm::data_model::permission::Token
+            ) -> ::core::result::Result<Self, Self::Error> {
+                if token.definition_id() !=
+                    &<Self as::iroha_wasm::validator::traits::Token>::definition_id()
+                {
+                    return Err(::iroha_wasm::validator::PermissionTokenConversionError::Id(
+                        token.definition_id().clone()
+                    ));
+                }
+
+                Ok(Self {
+                    #(#field_initializers),*
+                })
+            }
+        }
+    }
+}
+
+#[allow(clippy::arithmetic_side_effects)] // Triggers on quote! side
```

Comment:
Actually, can we fix the lint upstream? What happens is *not* arithmetic, and even if it were, it's not *our* problem. I'd keep it as is, create an issue here https://github.com/dtolnay/quote

Reply:
I created rust-lang/rust-clippy#10417, because [it seems like](https://github.com/dtolnay/quote/issues/232#issuecomment-1296451876) it should be fixed on `clippy` side